plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java

```diff
@@ -0,0 +1,258 @@
+     */
+    public abstract String stringInLogExpected();
+
+    @Before
+    public void saveJenkinsConfig() throws IOException {
+        File jenkinsConfig = new File(r.home, ConfigurationAsCode.DEFAULT_JENKINS_YAML_PATH);
+        if (jenkinsConfig.exists()) {
+            jenkinsConfigContentOld = FileUtils.readFileToString(jenkinsConfig);
+        }
+
+        cascJenkinsConfigPropOld = System.getProperty("casc.jenkins.config", null);
+    }
+
+    /*
+    Annotated with @After is executed before the second step is ended.
+    We need to call it at the end of the second step.
+     */
+    //@After
+    public void restoreJenkinsConfig() throws IOException {
+        if (jenkinsConfigContentOld != null) {
+            writeToFile(jenkinsConfigContentOld, new File(r.home, ConfigurationAsCode.DEFAULT_JENKINS_YAML_PATH).getAbsolutePath());
+        } else {
+            //We remove the file created during the test, it may affect other tests, for example, if the jenkins.yaml
+            //is invalid
+            Files.deleteIfExists(new File(r.home, ConfigurationAsCode.DEFAULT_JENKINS_YAML_PATH).toPath());
+        }
+
+        if (cascJenkinsConfigPropOld != null) {
+            System.setProperty("casc.jenkins.config", cascJenkinsConfigPropOld);
+        } else {
+            System.clearProperty("casc.jenkins.config");
+        }
+    }
+
+    /**
+     * 1.  Configure the instance with the {@link #configResource()} implemented.
+     * 2.  Check it was configured correctly.
+     * 3.  Export the full Jenkins configuration via Web UI (commented as it fails with maven defaultProperties).
+     * 4.  Export the schema via Web UI (commented out as there is no good validator so far).
+     * 5.  Verify the Jenkins configuration against the schema (commented out as there is no good validator so far).
+     * 6.  Check the Jenkins configuration is valid via Web UI (used the plugin config so far).
+     * 7.  Apply the Jenkins configuration via Web UI (maybe not needed, but we test it as well).
+     * 8.  Write the Jenkins configuration to $JENKINS_ROOT/jenkins.yaml.
+     * 9.  Restart Jenkins.
+     * 10. Check the {@link #stringInLogExpected()} is set during the restart.
+     *
+     * @throws IOException If an exception is thrown managing resources or files.
+     */
+    @Test
+    public void exportImportRoundTrip() throws IOException {
+        String resourcePath = configResource();
+        String resourceContent = getResourceContent(resourcePath);
+
+        assertNotNull(resourcePath);
+        assertNotNull(resourceContent);
+
+        r.then(step -> {
+            // Configure and validate
+            configureWithResource(resourcePath);
+            assertConfiguredAsExpected(r, resourceContent);
+
+            // Get the full configuration
+            //String jenkinsConf = getJenkinsConfViaWebUI();
+            //hack: the full Jenkins config fails due to defaultProperties of maven, we use the config of the plugin
```

Comment:
there's a PR for it but seems unlikely to be merged ☹️
may need to hack something into jcasc to blacklist it
https://github.com/jenkinsci/jenkins/pull/3992

Reply:
I've added a workaround, but would it be great if it works. We then can do the full export / replace roundtrip :smiley:

Reply:
Without that working I'm not that comfortable with this PR, i.e if a plugin user implements this test then there's no coverage for the exporting of their plugin.

The the bit which seems to be missing is without asserting the export is what you expect you risk the plugin author:
1. not viewing the export and realising it's not what they expect
2. unintentionally breaking the config in an update but not realising because the round trip works

Reply:
Hi Tim. I've tried to check the exported configuration has the same values as the YAML used. But, again, I faced a bug and I've commented out the code.

When running Jenkins normally:
You can add secrets ciphered in the YAML. CasC uploads it without reciphering. And when you export the configuration, the data is the same as it was in the yaml file.

When running with mvn running tests or hpi:run:
CasC ciphers the secrets, no matter if they were already ciphered in the yaml.

This bug makes impossible for plugins with secrets in the yaml to pass the test. We can solve this issue to uncomment this hack, or fix the full export to uncomment both hacks I did. :-)

Anyway, I've pushed the code.

Reply:
I've added the RoundTripMailerTest to prove it's working fine (when no secrets in the yaml)